threads.go

```diff
@@ -0,0 +1,357 @@
+package threads
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	bserv "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/textileio/go-textile-core/crypto"
+	"github.com/textileio/go-textile-core/thread"
+	tserv "github.com/textileio/go-textile-core/threadservice"
+	tstore "github.com/textileio/go-textile-core/threadstore"
+	"github.com/textileio/go-textile-threads/cbor"
+	pb "github.com/textileio/go-textile-threads/pb"
+	"github.com/textileio/go-textile-threads/util"
+	logger "github.com/whyrusleeping/go-logging"
+	"google.golang.org/grpc"
+)
+
+var log = logging.Logger("threads")
+
+const (
+	IPEL                     = "ipel"
+	IPELCode                 = 406
+	IPELVersion              = "0.0.1"
+	IPELProtocol protocol.ID = "/" + IPEL + "/" + IPELVersion
+)
+
+var addrProtocol = ma.Protocol{
+	Name:       IPEL,
+	Code:       IPELCode,
+	VCode:      ma.CodeToVarint(IPELCode),
+	Size:       ma.LengthPrefixedVarSize,
+	Transcoder: ma.TranscoderP2P,
+}
+
+func init() {
+	if err := ma.AddProtocol(addrProtocol); err != nil {
+		panic(err)
+	}
+}
+
+type threads struct {
+	host       host.Host
+	blocks     bserv.BlockService
+	service    *service
+	dagService format.DAGService
+	ctx        context.Context
+	cancel     context.CancelFunc
+	tstore.Threadstore
+}
+
+func NewThreads(ctx context.Context, h host.Host, bs bserv.BlockService, ds format.DAGService, ts tstore.Threadstore, debug bool) (tserv.Threadservice, error) {
+	var err error
+	if debug {
+		err = setLogLevels(map[string]logger.Level{
+			"threads":     logger.DEBUG,
+			"threadstore": logger.DEBUG,
+		}, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	t := &threads{
+		host:        h,
+		blocks:      bs,
+		dagService:  ds,
+		ctx:         ctx,
+		cancel:      cancel,
+		Threadstore: ts,
+	}
+	t.service, err = newService(t)
+	if err != nil {
+		return nil, err
+	}
+
+	listener, err := gostream.Listen(h, IPELProtocol)
```

Comment:
gostream returns a `net.Listener` which wraps our host. This can be used just like any other listener!